homeassistant/components/blue_current/__init__.py

```diff
@@ -0,0 +1,193 @@
+            """Loop over the charge points and get their data."""
+            for entry in data:
+                evse_id = entry[EVSE_ID]
+                model = entry[MODEL_TYPE]
+                name = entry[ATTR_NAME]
+                self.add_charge_point(evse_id, model, name)
+                await self.get_charge_point_data(evse_id)
+            await self.client.get_grid_status(data[0][EVSE_ID])
+
+        object_name: str = message[OBJECT]
+
+        # gets charge point ids
+        if object_name == CHARGE_POINTS:
+            charge_points_data: list = message[DATA]
+            await handle_charge_points(charge_points_data)
+
+        # gets charge point key / values
+        elif object_name in VALUE_TYPES:
+            value_data: dict = message[DATA]
+            evse_id = value_data.pop(EVSE_ID)
+            self.update_charge_point(evse_id, value_data)
+
+        # gets grid key / values
+        elif GRID in object_name:
+            data: dict = message[DATA]
+            self.grid = data
+            self.dispatch_grid_update_signal()
+
+    async def get_charge_point_data(self, evse_id: str) -> None:
+        """Get all the data of a charge point."""
+        await self.client.get_status(evse_id)
+
+    def add_charge_point(self, evse_id: str, model: str, name: str) -> None:
+        """Add a charge point to charge_points."""
+        self.charge_points[evse_id] = {MODEL_TYPE: model, ATTR_NAME: name}
+
+    def update_charge_point(self, evse_id: str, data: dict) -> None:
+        """Update the charge point data."""
+        self.charge_points[evse_id].update(data)
+        self.dispatch_value_update_signal(evse_id)
+
+    def dispatch_value_update_signal(self, evse_id: str) -> None:
+        """Dispatch a value signal."""
+        async_dispatcher_send(self.hass, f"{DOMAIN}_value_update_{evse_id}")
+
+    def dispatch_grid_update_signal(self) -> None:
+        """Dispatch a grid signal."""
+        async_dispatcher_send(self.hass, f"{DOMAIN}_grid_update")
+
+    async def start_loop(self) -> None:
+        """Start the receive loop."""
+        try:
+            await self.client.start_loop(self.on_data)
+        except BlueCurrentException as err:
+            LOGGER.warning(
+                "Disconnected from the Blue Current websocket. Retrying to connect in background. %s",
+                err,
+            )
+
+            async_call_later(self.hass, SMALL_DELAY, self.reconnect)
+
+    async def reconnect(self, _event_time: datetime | None = None) -> None:
+        """Keep trying to reconnect to the websocket."""
+        try:
+            await self.connect(self.config.data[CONF_API_TOKEN])
+            LOGGER.info("Reconnected to the Blue Current websocket")
+            self.hass.loop.create_task(self.start_loop())
+            await self.client.get_charge_points()
+        except RequestLimitReached:
+            set_entities_unavalible(self.hass, self.config.entry_id)
```

Comment:
Since connector is inserted into the sensors, wouldn't it make more sense to add an `available` property to the entities where they can maybe return a `connector.available`?

Reply:
Like it would be a more _natural_ way of handling unavailable, because setting entities unavailable via the registry is not common